src/source.cpp

```diff
@@ -79,6 +93,146 @@ unique_ptr<Source> Source::create(pugi::xml_node node)
   }
 }
 
+void Source::read_constraints(pugi::xml_node node)
+{
+  // Check for constraints node. For backwards compatibility, if no constraints
+  // node is given, still try searching for domain constraints from top-level
+  // node.
+  pugi::xml_node constraints_node = node.child("constraints");
+  if (constraints_node) {
+    node = constraints_node;
+  }
+
+  // Check for domains to reject from
+  if (check_for_node(node, "domain_type")) {
+    std::string domain_type = get_node_value(node, "domain_type");
+    if (domain_type == "cell") {
+      domain_type_ = DomainType::CELL;
+    } else if (domain_type == "material") {
+      domain_type_ = DomainType::MATERIAL;
+    } else if (domain_type == "universe") {
+      domain_type_ = DomainType::UNIVERSE;
+    } else {
+      fatal_error(
+        std::string("Unrecognized domain type for constraint: " + domain_type));
+    }
+
+    auto ids = get_node_array<int>(node, "domain_ids");
+    domain_ids_.insert(ids.begin(), ids.end());
+  }
+
+  if (check_for_node(node, "time_bounds")) {
+    auto ids = get_node_array<double>(node, "time_bounds");
+    if (ids.size() != 2) {
+      fatal_error("Time bounds must be represented by two numbers.");
+    }
+    time_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+  if (check_for_node(node, "energy_bounds")) {
+    auto ids = get_node_array<double>(node, "energy_bounds");
+    if (ids.size() != 2) {
+      fatal_error("Energy bounds must be represented by two numbers.");
+    }
+    energy_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+
+  if (check_for_node(node, "fissionable")) {
+    only_fissionable_ = get_node_value_bool(node, "fissionable");
+  }
+
+  // Check for how to handle rejected particles
+  if (check_for_node(node, "rejection_strategy")) {
+    std::string rejection_strategy = get_node_value(node, "rejection_strategy");
+    if (rejection_strategy == "kill") {
+      rejection_strategy_ = RejectionStrategy::KILL;
+    } else if (rejection_strategy == "resample") {
+      rejection_strategy_ = RejectionStrategy::RESAMPLE;
+    } else {
+      fatal_error(std::string(
+        "Unrecognized strategy source rejection: " + rejection_strategy));
+    }
+  }
+}
+
+SourceSite Source::sample_with_constraints(uint64_t* seed) const
+{
+  bool accepted = false;
+  SourceSite site;
+
+  while (!accepted) {
```

Comment:
Should we warn on an acceptance fraction less than 5% or so?

Reply:
A matter of taste but I'd say yes. It's the prior default behavior, and it tends to signify a mistake, regardless of strategy.

Reply:
I also agree -- I implemented this on the branch already.